akka-projection-grpc/src/main/scala/akka/projection/grpc/consumer/scaladsl/GrpcReadJournal.scala

```diff
@@ -262,51 +267,55 @@ final class GrpcReadJournal private (
-      .collect {
-        case ConsumerFilter.FilterCommand(`streamId`, criteria) =>
-          val protoCriteria = criteria.map {
-            case ConsumerFilter.IncludeEntityIds(entityOffsets) =>
-              FilterCriteria(FilterCriteria.Message.IncludeEntityIds(IncludeEntityIds(entityOffsets.map {
-                case ConsumerFilter.EntityIdOffset(entityId, seqNr) => EntityIdOffset(entityId, seqNr)
-              }.toVector)))
-            case ConsumerFilter.RemoveIncludeEntityIds(entityIds) =>
-              FilterCriteria(FilterCriteria.Message.RemoveIncludeEntityIds(RemoveIncludeEntityIds(entityIds.toVector)))
-            case ConsumerFilter.ExcludeEntityIds(entityIds) =>
-              FilterCriteria(FilterCriteria.Message.ExcludeEntityIds(ExcludeEntityIds(entityIds.toVector)))
-            case ConsumerFilter.RemoveExcludeEntityIds(entityIds) =>
-              FilterCriteria(FilterCriteria.Message.RemoveExcludeEntityIds(RemoveExcludeEntityIds(entityIds.toVector)))
-            case ConsumerFilter.ExcludeRegexEntityIds(matching) =>
-              FilterCriteria(FilterCriteria.Message.ExcludeMatchingEntityIds(ExcludeRegexEntityIds(matching.toVector)))
-            case ConsumerFilter.RemoveExcludeRegexEntityIds(matching) =>
-              FilterCriteria(
-                FilterCriteria.Message.RemoveExcludeMatchingEntityIds(RemoveExcludeRegexEntityIds(matching.toVector)))
-          }
+    val consumerFilter = ConsumerFilter(typedSystem)
+
+    def inReqSource(initCriteria: immutable.Seq[ConsumerFilter.FilterCriteria]): Source[StreamIn, NotUsed] =
+      Source
+        .actorRef[ConsumerFilter.SubscriberCommand](
+          completionMatcher = PartialFunction.empty,
+          failureMatcher = PartialFunction.empty,
+          bufferSize = 1024,
+          OverflowStrategy.fail)
+        .collect {
+          case ConsumerFilter.UpdateFilter(`streamId`, criteria) =>
+            val protoCriteria = toProtoFilterCriteria(criteria)
+            StreamIn(StreamIn.Message.Filter(FilterReq(protoCriteria)))
+
+          case ConsumerFilter.Replay(`streamId`, entityOffsets) =>
+            val protoEntityOffsets = entityOffsets.map {
+              case ConsumerFilter.EntityIdOffset(entityId, seqNr) => EntityIdOffset(entityId, seqNr)
+            }.toVector
+            StreamIn(StreamIn.Message.Replay(ReplayReq(protoEntityOffsets)))
+        }
+        .mapMaterializedValue { ref =>
+          consumerFilter.ref ! ConsumerFilter.Subscribe(streamId, initCriteria, ref)
+          NotUsed
+        }
 
-          StreamIn(StreamIn.Message.Filter(FilterReq(protoCriteria)))
+    val initFilter = {
+      import akka.actor.typed.scaladsl.AskPattern._
+      import scala.concurrent.duration._
+      implicit val scheduler: Scheduler = typedSystem.scheduler
+      implicit val askTimeout: Timeout = 10.seconds // FIXME config
+      consumerFilter.ref.ask[ConsumerFilter.CurrentFilter](ConsumerFilter.GetFilter(streamId, _))
```

Comment:
Yeah, the setup is somewhat difficult. We need the current filter first, to include in the grpc InitReq. Then we register the subscriber when we have the Source.actorRef. Do you see a way to make it in one go, like GetAndSubscribe I'm open for suggestions.

I think we have to include it in the InitReq and not send it right after as an update FilterReq, because then it could already have emitted some events from the producer before it receives the FilterReq.

Reply:
No obvious way no, going separate actorrefs in the subscribe requests for initial state and updates would perhaps be possible, but still messy on this side to piece together, so probably not worth it.